src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Ascii.Transcoding.cs

```diff
@@ -0,0 +1,86 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace System.Buffers.Text
+{
+    public static partial class Ascii
+    {
+        /// <summary>
+        /// Copies text from a source buffer to a destination buffer, converting
+        /// from ASCII to UTF-16 during the copy.
+        /// </summary>
+        /// <param name="source">The source buffer from which ASCII text is read.</param>
+        /// <param name="destination">The destination buffer to which UTF-16 text is written.</param>
+        /// <param name="bytesConsumed">The number of bytes actually read from <paramref name="source"/>.</param>
+        /// <param name="charsWritten">The number of chars actually written to <paramref name="destination"/>.</param>
+        /// <returns>An <see cref="OperationStatus"/> describing the result of the operation.</returns>
+        public static unsafe OperationStatus ToUtf16(ReadOnlySpan<byte> source, Span<char> destination, out int bytesConsumed, out int charsWritten)
+        {
+            nuint numElementsToConvert;
+            OperationStatus statusToReturnOnSuccess;
+
+            if (source.Length <= destination.Length)
+            {
+                numElementsToConvert = (uint)source.Length;
+                statusToReturnOnSuccess = OperationStatus.Done;
+            }
+            else
+            {
+                numElementsToConvert = (uint)destination.Length;
+                statusToReturnOnSuccess = OperationStatus.DestinationTooSmall;
+            }
+
+            fixed (byte* pSource = &MemoryMarshal.GetReference(source))
+            fixed (char* pDestination = &MemoryMarshal.GetReference(destination))
+            {
+                nuint numElementsActuallyConverted = ASCIIUtility.WidenAsciiToUtf16(pSource, pDestination, numElementsToConvert);
```

Comment:
Do we still need ASCIIUtility, or can the contents of that type be folded into Ascii?

Reply:
I would love to fold it and also get rid of all the pinning and use managed references (`ref`) everywhere.

@stephentoub Is there any chance you could review https://github.com/dotnet/runtime/pull/73055? Once I merge it I could fold it and avoid merge conflict.